jaxopt/_src/osqp.py

```diff
@@ -0,0 +1,474 @@
+    verbose: If verbose=1, print error at each iteration. If verbose=2, also print stepsizes and primal/dual variables.
+      Warning: verbose>0 will automatically disable jit.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+
+  [1] Stellato, B., Banjac, G., Goulart, P., Bemporad, A. and Boyd, S., 2020.
+  OSQP: An operator splitting solver for quadratic programs.
+  Mathematical Programming Computation, 12(4), pp.637-672.
+
+  [2] Boyd, S., Parikh, N., Chu, E., Peleato, B. and Eckstein, J., 2010.
+  Distributed Optimization and Statistical Learning via the Alternating Direction Method of Multipliers.
+  Machine Learning, 3(1), pp.1-122.
+  """
+  matvec_P: Optional[Callable] = None
+  matvec_A: Optional[Callable] = None
+  check_primal_dual_infeasability: bool = False
+  sigma: float = 1e-6
+  alpha: float = 1.6
+  eq_qp_solve: Callable = solve_normal_cg
+  eq_qp_solve_tol: Optional[float] = 1e-7  # relative tolerance. TODO: should it be absolute tolerance ?
+  eq_qp_solve_maxiter: Optional[int] = None
+  rho_start: float = 0.1
+  rho_eq_over_rho_ineq: float = 1e3
+  rho_min: float = 1e-6
+  rho_max: float = 1e6
+  rho_tol: float = 1e-4
+  tol_primal_unfeasible: float = 1e-4  # relative tolerance
+  tol_dual_unfeasible: float = 1e-4  # relative tolerance
+  maxiter: int = 4000
+  tol: float = 1e-3  # absolute tolerance. TODO: should it be absolute tolerance ?
+  verbose: int = 0
+  implicit_diff: bool = True
+  implicit_diff_solve: Optional[Callable] = None
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+
+  # class attributes (ignored by @dataclass)
+  UNSOLVED          = 0  # stopping criterion not reached yet
+  SOLVED            = 1  # feasible solution found with satisfying precision
+  DUAL_INFEASIBLE   = 2  # infeasible dual
+  PRIMAL_INFEASIBLE = 3  # infeasible primal
+  
+
+  def _eq_constraints(self, box):
+    l, u = box
+    is_eq = lambda li, ui: (ui-li) < self.rho_tol
+    return tree_map(is_eq, l, u)
+
+  def init_state(self, init_params, params_obj, params_eq, params_ineq):
+    x, z = init_params.primal
+    y    = init_params.dual_eq  # good approximation for nu in Equality constrained QP
+    rho  = tree_scalar_mul(self.rho_start, tree_ones_like(z))
+    coef = tree_where(self._eq_constraints(params_ineq), self.rho_eq_over_rho_ineq, 1)
+    rho  = tree_mul(coef, rho)
+    return OSQPState(iter_num=0,
+                     error=jnp.inf,
+                     status=OSQP.UNSOLVED,
+                     primal_residuals=jnp.inf,
+                     dual_residuals=jnp.inf,
+                     rho=rho,
+                     eq_qp_last_sol=(x, y))
+  
+  def init_params(self, params_obj, params_eq, params_ineq, x0=None):
```

Comment:
The user can warm start the algorithm with a prior on x0: `prior_kkt_solution = osqp.init_params(x0=prior_x0, **hyper_params)`.
  
If the user has a prior on all variables (including dual ones), it can warm start the algorithm by callling `run(init_params=prior_kkt_solution, **hyper_params)`.

Ideally, the `run()` would also support `run(init_params=None, **hyper_params)` that would fall back to call `init_params = osqp.init_params(x0=None, **hyper_params)` internally.

It might be overkill to proposer all those options. If you have any hint on which would be the best we can go for it !

On one hand, as the problen is convex, the starting point is unimportant. On the other hand, warm starting the algorithm within feasible set or close to the minimum (particularly useful if we want to solve a sequence of similar optimizations problems) it could dramatically speed up convergence.

Reply:
I think `init_params(self, init_x, params_obj, params_eq, params_ineq)` would be more consistent.